src/base/default_coupling.C

```diff
@@ -160,10 +160,8 @@ void DefaultCoupling::operator()
               active_neighbors.push_back(neigh);
 #endif
 
-              for (std::size_t a=0; a != active_neighbors.size(); ++a)
+              for (const auto & neighbor : active_neighbors)
```

Comment:
At some point I want to look at ```g++ -S``` output and verify that we don't actually lose anything by using a reference-to-pointer rather than copy-of-pointer in these range for loops over containers of ```Elem *```.  In theory I could imagine the compiler deciding that it can't prove the container's contents aren't being changed by function calls inside the loop and so it needs to do a double-dereference instead of a single for later Elem accesses.

Reply:
For the following simple test code:
```C++
#include <vector>

int main()
{
  std::vector<int *> v = {new int(1), new int(2)};
  int sum = 0;
  for (auto & ptr : v)
    sum += *ptr;
  
  return sum;
}
```
there is no difference in the generated assembly regardless of the `&` when compiling with `-O3` with g++ (GCC) 7.3.1 20180902.

Reply:
Oh, there's definitely not going to be a difference there.  But create a function that's in a separate translation unit and that has access to v, call it inside the loop, and see if *that* makes a difference?

Reply:
No difference in assembly for this simple case either...
```C++
#include <vector>

// Function defined in different TU. Note: the function's not really
// defined anywhere, as we are just generating assembly.
int f(int x);

int main()
{
  std::vector<int *> v = {new int(1), new int(2)};
  int sum = 0;
  for (auto & ptr : v)
    sum += f(*ptr);
  
  return sum;
}
```

Reply:
Looks like my theory was correct; you just have to get a little more complicated to show it:
```
#include <vector>

// Function defined in different TU. Note: the function's not really
// defined anywhere, as we are just generating assembly.
template <typename T>
int f(T & v, int x);

int main()
{
  std::vector<int *> v = {new int(1), new int(2)};
  int sum = 0;
  for (auto & ptr : v)
    {
      sum += f(v, *ptr);
      sum += f(v, *ptr);
    }

  return sum;
}
```

If you leave the reference in the loop iterator declaration, then (with gcc 8.1 defaults at least) the assembly code does a double-dereference before each use of ```*ptr```; if you remove the reference then a single-dereference gets pulled up to the top of the loop body, for one fewer dereference per ```*ptr```-use-after-the-first per loop iteration.

If I turn on -O3 with gcc 8.1 I can't make head or tail of the assembly anymore, but (after supplying a "return 1" stub f in a separate file and making the loops longer) callgrind seems to think a loop with N iterations over M ```ptr``` uses each entails ```(M-1)*N-2``` more Ir (instruction read) events with the reference in the code.  (Unless M is 1 or 0, in which case whatever that "-2" is goes away and the codes are identical)